### PR TITLE
Perf Report

### PR DESCRIPTION
**Pull request by PathikritGhosh**
- Fix the perf report values traced. I have changed how few of the fields are computed and added computations for some fields that were not being computer earlier.
- Add functionality to parse perf report and show a subset if fields based on the decision tree.

### CODE REVIEW

File: src/ck-pics/picstreenode.C [hunk 1]
```diff
@@ -214,6 +234,12 @@ bool Condition::test(double *input) {
   return ret;
 }
 
+/**
+ * PICS suggested solutions.
+ * Applied to code iff PICS is used for tuning otherwise just prints a suggestion
```
Comment (rbuch) [inline src/ck-pics/picstreenode.C:239]: This function never applies a solution to code, does it? I think this comment is a little misleading, it should clearly describe the behavior of just _this function_, along with some other context if necessary. Where do the solutions get applied?

File: src/ck-pics/trace-perf.C [hunk 2]
```diff
@@ -307,14 +311,23 @@ void TraceAutoPerf::traceClose(void)
...
 
-void TraceAutoPerf::summarizeObjectInfo(double &maxtime, double &totaltime,
+void TraceAutoPerf::summarizeObjectInfo(double &maxtime, double &totaltime, double &mintime,
     double &maxMsgCount, double &totalMsgCount, double &maxMsgSize,
-    double &totalMsgSize, double &numObjs) {
+    double &totalMsgSize, double &numObjs, double &maxBytesPerMsg, double &minBytesPerMsg) {
   void *maximum = NULL;
+
+  if(objectLoads.size() > 0) {
+    mintime = objectLoads.begin()->second->executeTime;
```
Comment (rbuch) [inline src/ck-pics/trace-perf.C:320]: Should this also set `maxtime` or is that guaranteed to be initialized to 0 or something?

File: src/ck-pics/trace-perf.C [hunk 3]
```diff
@@ -328,43 +341,54 @@ void TraceAutoPerf::summarizeObjectInfo(double &maxtime, double &totaltime,
...
-  currentSummary->data[MIN_AppPercentage] = currentSummary->data[MAX_AppPercentage] = appTime();
-  currentSummary->data[AVG_AppPercentage] = appTime();
-  currentSummary->data[AVG_TuningOverhead] = tuneOverheadTotalTime; 
-  currentSummary->data[MIN_OverheadPercentage] = currentSummary->data[MAX_OverheadPercentage] = overheadTime(); 
-  currentSummary->data[AVG_OverheadPercentage] = overheadTime()/currentSummary->data[AVG_TotalTime];
+  currentSummary->data[AVG_TotalTime] = totalTraceTime();
+  currentSummary->data[MIN_IdlePercentage] = currentSummary->data[MAX_IdlePercentage] = currentSummary->data[AVG_IdlePercentage] = idleRatio();
+  currentSummary->data[MIN_OverheadPercentage] = currentSummary->data[MAX_OverheadPercentage] = currentSummary->data[AVG_OverheadPercentage] = overheadRatio();
+  currentSummary->data[MIN_UtilizationPercentage] = currentSummary->data[MAX_UtilizationPercentage] = currentSummary->data[AVG_UtilizationPercentage] = utilRatio();
+  currentSummary->data[MAX_NumInvocations] = currentSummary->data[AVG_NumInvocations] = currentSummary->data[AVG_NumInvocations] = (double)totalEntryMethodInvocations;
```
Comment (rbuch) [inline src/ck-pics/trace-perf.C:362]: This seems to set `AVG_NumInvocations` twice.

File: src/ck-pics/trace-perf.C [hunk 4]
```diff
@@ -328,43 +341,54 @@ void TraceAutoPerf::summarizeObjectInfo(double &maxtime, double &totaltime,
...
+
+  summarizeObjectInfo(currentSummary->data[MAX_LoadPerObject], currentSummary->data[AVG_LoadPerObject], currentSummary->data[MIN_LoadPerObject],
+                      currentSummary->data[MAX_NumMsgsPerObject],  currentSummary->data[AVG_NumMsgsPerObject],
+                      currentSummary->data[MAX_BytesPerObject], currentSummary->data[AVG_BytesPerObject], currentSummary->data[AVG_NumObjectsPerPE],
+                      currentSummary->data[MAX_BytesPerMsg], currentSummary->data[MIN_BytesPerMsg]);
+  currentSummary->data[AVG_NumMsgRecvPerPE] = currentSummary->data[AVG_NumMsgsPerObject];
+  currentSummary->data[AVG_BytesMsgRecvPerPE] = currentSummary->data[AVG_BytesPerObject];
```
Comment (rbuch) [inline src/ck-pics/trace-perf.C:369]: Why is this using the per object values to set the per PE values? Are they combined somehow somewhere else?

File: src/ck-pics/picsautoperf.C [hunk 5]
```diff
@@ -356,37 +361,36 @@ void TraceAutoPerfBOC::setCbAndRun(bool fromGlobal, int fromPE, CkCallback cb) {
...
+  data[AVG_LoadPerPE] = data[AVG_UtilizationPercentage] * totaltime;
+  data[AVG_CacheMissRate] = data[AVG_CacheMissRate]/numpes;
+
+  data[AVG_NumMsgsPerObject] /= (data[AVG_NumObjectsPerPE]*numpes);
+  data[AVG_BytesPerObject] /= (data[AVG_NumObjectsPerPE]*numpes);
+  data[AVG_LoadPerObject] /= (data[AVG_NumObjectsPerPE]*numpes);
+  data[AVG_BytesPerMsg] = data[AVG_BytesPerObject]/data[AVG_NumMsgsPerObject];
+
+  data[AVG_NumMsgPerPE] = data[AVG_NumMsgsPerObject]/numpes;
+  data[AVG_BytesPerPE] = data[AVG_BytesPerObject]/numpes;
```
Comment (rbuch) [inline src/ck-pics/picsautoperf.C:384]: Should these be `AVG_xPerObject *  AVG_NumObjectsPerPE`?

File: src/ck-pics/trace-perf.C (hunk 4 above)
Comment (PathikritGhosh) [reply src/ck-pics/trace-perf.C:369]: My bad. This is an example of using incorrect variable names. currentSummary->data[AVG_NumMsgsPerObject] is the AVG_NumMsgRecvPerPE, which is later converted to Avg_NumMsgPerObject. Same for the second one. Also leads to the comments you mentioned https://github.com/UIUC-PPL/charm/pull/3419#discussion_r674142593 here.
Will fix this in the next commit.

File: src/ck-pics/picstreenode.C [hunk 6]
```diff
@@ -110,33 +115,48 @@ void Condition::printMe() {
...
+    fprintf(fp, "%s is too low.\n", str.substr(pos+1).c_str());
+  } else if(strstr(str.c_str(), "High")) {
+    fprintf(fp, "%s is too high.\n", str.substr(pos+1).c_str());
+  } else if(str.c_str(), "Small") {
+    fprintf(fp, "%s is too small.\n", str.substr(pos+1).c_str());
+  } else if(str.c_str(), "Many") {
+    fprintf(fp, "%s is too many.\n", str.substr(pos+1).c_str());
+  } else if(str.c_str(), "Few") {
+    fprintf(fp, "%s is too few.\n", str.substr(pos+1).c_str());
+  } else if(str.c_str(), "Long") {
```
[the remark this answers is not in the document]
Comment (PathikritGhosh) [reply src/ck-pics/picstreenode.C]: Made this change.

**Pull request closed without merging**